File: typescript/patterns-use-cases/src/sagas/activities/flight_service.ts
import * as restate from "@restatedev/restate-sdk";

export const flightsService = restate.service({
    name: "FlightsService",
    handlers: {
      reserve: async (ctx: restate.Context, _req: { flightId: string, passengerName: string }) => {
        const flightBookingId = ctx.rand.uuidv4();
        console.info(`Flight ${flightBookingId} reserved`);
        return flightBookingId;
      },

      confirm: async (ctx: restate.Context, req: {flightBookingId: string}) => {
        console.info(`Flight ${req.flightBookingId} confirmed`);
      },

      cancel: async (ctx: restate.Context, req: {flightBookingId: string}) => {
        console.info(`Flight ${req.flightBookingId} cancelled`);
      },
    }
});

export type FlightsService = typeof flightsService;

File: go/patterns-use-cases/src/sagas/activities/flightservice.go
package activities

import (
	restate "github.com/restatedev/sdk-go"
	"log/slog"
)

type FlightBookingRequest struct {
	FlightId      string `json:"flight_id"`
	PassengerName string `json:"passenger_name"`
}

type FlightService struct{}

func (FlightService) Reserve(ctx restate.Context, _flightBookingRequest FlightBookingRequest) (string, error) {
	flightBookingId := restate.Rand(ctx).UUID().String()
	slog.Info("Flight reserved: " + flightBookingId)
	return flightBookingId, nil
}

func (FlightService) Confirm(ctx restate.Context, flightBookingId string) error {
	slog.Info("Flight confirmed: " + flightBookingId)
	return nil
}

func (FlightService) Cancel(ctx restate.Context, flightBookingId string) error {
	slog.Info("Flight cancelled: " + flightBookingId)
	return nil
}
